# Example Usage - `py_atmosphere`

This document displays how to use `py_atmosphere` to extract atmospheric properties from the package, transform units and calculate Mach number for an object moving through stagnant air:

## Import Package Functions

In [1]:
from py_atmosphere.py_atmosphere import py_atmosphere
from py_atmosphere.mach_number import mach_number
from py_atmosphere.temp_conversion import temp_conversion

## Technical Background

This package contains the mathematical representation of NASA's Glenn Research Center [Earth Atmosphere Model](https://www.grc.nasa.gov/www/k-12/airplane/atmosmet.html#:~:text=In%20the%20troposphere%2C%20the%20temperature,atmosphere%20model%20is%20also%20available), which is a simplified representation of the International Standard Atmosphere.

Under this context, the Earth's atmosphere is considered a sheet of air extending from the Earth's surface to the edge of space. Within the atmosphere, very complex chemical, thermodynamic and fluid dynamics effects occur. Given these effects, the properties of the air are constantly changing with time and place.

These effects have the following instintive impacts:
- Given the effect that the sun has on the surface of the Earth, and how this heat is diffused up through the atmosphere, the air temperature is highest near the surface of the planet and decreases with increasing altitude.
- The pressure of the air is related to the weight of the air over a given location. As we increase altitude through the atmosphere, the amount of air above us is less. For this reason, air pressure decreases as altitude increases.

These notions are incredibly important in the Aerospace industry, as how the atmosphere and the air properties change can have a direct impact in Aircraft operation throughout the flight envelope of an aircraft, for example. To help aircraft designers, it is useful to have a Standard Atmosphere model of the variation of properties throughout the atmosphere. 

## The `py_atmosphere()` function

First, we need to give a brief introduction to the modeled layers of the atmosphere.

- From 0 to 11,000 meters, we have the Troposphere.
- From 11,000 meters to 25,000 meters, we have the Lower Stratosphere.
- Beyond 25,000 meters, we have the Upper Stratosphere

Note: This implementation of the model is restricted to an input altitude of 0m to 25,000 m. Higher altitudes require more complex modeling to avoid significant inaccuracies, which are out of the scope of this package. The selected range is sufficient when considering that the average altitude during cruise flight in civil aircraft is between 10,000 and 12,000 meters. 

The standard model consists of the following equations to calculate Temperature and Pressure given an Altitude.

Variables:

- T = Temperature (°C - degrees Celsius)
- P = Pressure (K-Pa - KiloPascal)
- h = Altitude (m - meters)

In the Troposphere, for $h < 11,000m$:

$T = 15.04 - 0.00649*h$

$P = 101.29 * [{\frac{T + 273.1}{288.08}}]^{5.256}$

In the Lower Stratosphere, for $11,000m < h < 25,000m$:

$T = -56.46$

$P = 22.65 * e^{1.73-0.000157*h}$

For the Upper Stratosphere, for $h > 25,000m$:

$T = -131.21 + 0.00299*h$

$P = 2.488 * [{\frac{T + 273.1}{216.6}}]^{-11.388}$

Considering these equations, we can use the `py_atmosphere` function to calculate the standard pressure and temperature given an altitude. Depending on the inquired altitude, the package will select the appropriate set of equations to calculate Temperature and Pressure.

Note that the inquired altitude may be entered in different units:

- `m` for meters
- `km` for kilometers
- `feet` for feet
- `miles` for miles

Example: Calculate the Pressure and Temperature at an altitude of 0 meters, 10 kilometers, 10,000 ft and 15 miles:

In [2]:
py_atmosphere(0, "m")

[0, 15.04, 101.29]

How to read the output: At an altitude of 0 meters, the standard atmosphere calculates an air Temperature of 15°C and air Pressure of 101.29 kiloPascals.

In [3]:
py_atmosphere(10, "km")

[10000, -49.86000000000001, 26.547435615976987]

How to read the output: At an altitude of 10 kilometers, the standard atmosphere calculates an air Temperature of -49.86°C and air Pressure of 26.54 kiloPascals. Note that the output converts entered altitude to meters. In this case, 1 kilometer is equal to 10,000 meters.

In [4]:
py_atmosphere(10000, "ft")

[3048.0, -4.741520000000001, 69.84232244791907]

How to read the output: At an altitude of 10,000 feet, the standard atmosphere calculates an air Temperature of -4.74°C and air Pressure of 69.84 kiloPascals. Note that the output converts entered altitude to meters. In this case, 10,000 feet is equal to 3,048 meters.

In [5]:
py_atmosphere(15, "miles")

[24140.1, -56.46, 2.886844831523778]

How to read the output: At an altitude of 15 miles, the standard atmosphere calculates an air Temperature of -56.46°C and air Pressure of 2.88 kiloPascals. Note that the output converts entered altitude to meters. In this case, 15 miles is equal to 24,140 meters.

As mentioned, the function does not work for altitude outside of the range of 0 to 25,000 meters.

In [6]:
try:
    py_atmosphere(-1, "m")
except ValueError as e:
    print(f"Expected error: {e}. Altitude must be a non-negative value")

Expected error: Altitude must be a non-negative value. Altitude must be a non-negative value


In [7]:
try:
    py_atmosphere(25001, "m")
except ValueError as e:
    print(f"Expected error: {e}. Invalid altitude value. Model maximum value is 25,000 m.")

Expected error: Invalid altitude value. Model maximum value is 25,000 m.. Invalid altitude value. Model maximum value is 25,000 m.


## The `temp_conversion()` function

Given the frequent need to see these properties in different units, the `py_atmosphere` package contains a function to perform convenient conversions for altitude, temperature and pressure. The input and output of this function shares the same list structure as the output from the `py_atmosphere` function.

This function requires the user to select which property needs conversion and to which units. Note that the package can currently perform only one conversion at a time.

The conversion units accepted for each property are:

- Altitude: "m" for meters, "km" for kilometers, "ft" for feet, and "miles" for miles.
- Temperature: "C" for Celsius, "F" for Fahrenheit, "K" for Kelvin, and "R" for Rankine.
- Pressure: "kPa" for KiloPascal, "bar" for Pascal, and "psia" for pounds per square inch. 

### Temperature Conversions

In [8]:
example_temperature = py_atmosphere(0, "m")
example_temperature

[0, 15.04, 101.29]

In [9]:
temp_conversion(example_temperature, "temperature", "C")

[0, 15.04, 101.29]

In [10]:
temp_conversion(example_temperature, "temperature", "F")

[0, 59.071999999999996, 101.29]

In [11]:
temp_conversion(example_temperature, "temperature", "K")

[0, 288.19, 101.29]

In [12]:
temp_conversion(example_temperature, "temperature", "R")

[0, 518.742, 101.29]

### Pressure Conversions

In [13]:
example_pressure = py_atmosphere(0, "m")
example_pressure

[0, 15.04, 101.29]

In [14]:
temp_conversion(example_pressure, "pressure", "kPa")

[0, 15.04, 101.29]

In [15]:
temp_conversion(example_pressure, "pressure", "bar")

[0, 15.04, 1.0129000000000001]

In [16]:
temp_conversion(example_pressure, "pressure", "psia")

[0, 15.04, 14.690868633000001]

### Altitude Conversion

In [17]:
example_altitude = py_atmosphere(1000, "m")
example_altitude

[1000, 8.549999999999999, 90.04204320972651]

In [18]:
temp_conversion(example_altitude, "altitude", "m")

[1000, 8.549999999999999, 90.04204320972651]

In [19]:
temp_conversion(example_altitude, "altitude", "km")

[1.0, 8.549999999999999, 90.04204320972651]

In [20]:
temp_conversion(example_altitude, "altitude", "ft")

[3280.84, 8.549999999999999, 90.04204320972651]

In [21]:
temp_conversion(example_altitude, "altitude", "miles")

[0.621371, 8.549999999999999, 90.04204320972651]

## The `mach_number()` function

The `py_atmosphere` package is also prepared to calculate the Mach Number of a moving object in stagnant air.

Taking a step back, it is necessary to describe the concept of [Speed of Sound](https://www.grc.nasa.gov/www/BGH/sound.html). Speed of sound (`a`) is the speed in which sound (or formally, a sound wave) propagates though a medium. In an ideal gas (a common assumption in design activities that describes how gases beave), the speed of sound is only dependent on the medium's temperature and composition.

$a = \sqrt{\gamma RT}$

Where:

- Air property $\gamma$ - Ratio of specific heats - $1.4$ for air.
- Air property `R` - Gas constant - $287.05 J / kg * K$
- `T` - Air temperature in K

The Mach Number is dimensionless quantity in fluid dynamics that represents the ratio of flow velocity around a boundary relative to the local speed of sound. In aerospace applications, this ratio represents the relation between an aircraft's speed as it moves through stagnant air, where the local speed of sound is a property that is only function of the local temperature.

The Mach number is widely used throughout fluid mechanics applications as it is descriptive of the compressibility of the fluids. In the Aerospace industry it is also descriptive of subsonic, sonic and supersonic flight. Therefore, it is an invaluable design parameter.

$Ma = \frac{c}{a}$

Where:

- `a` - Speed of Sound in medium
- `c` - Speed of an object through same medium

The `mach_number` function takes the temperature calculated by the `py_atmosphere` function, and along with an input `speed` of an object, it calculates the Mach Number. Given that temperature is a function of altitude, we can see that for an object moving at a constant speed, the Mach number changes with altitude, as temperature changes as well. 

Under this consideration, the `mach_number` function receives an `altitude`, altitude `units` and object `speed` in meters per second. Note that just as the `py_atmosphere` function, `mach_number` also accepts altitude units in "m" for meters, "km" for kilometers, "ft" for feet, and "miles" for miles.

Example: 

What is the Mach Number of an object moving at 350 m/s?

Note: A Mach Number < 1 describes "subsonic" flight, a Mach Number = 1 describes "sonic" flight, and Mach Number > 1 describes "supersonic" flight.

First, at an altitude of 0 meters:

We can see that an object moving at 300 m/s at an altitude of 0 meters describes a "Subsonic" movement.

In [22]:
mach_number(0, "m", 300)

0.8815338791328668

We can see that an object moving at the same 300 m/s, but at an altitude of 10,000 meters, describes supersonic flight (in other words, it is breaking the sound barrier!)

In [23]:
mach_number(10000, "m", 300)

1.0014835599066014